executor/src/witgen/machines/block_machine.rs

```diff
@@ -295,8 +312,28 @@ impl<T: FieldElement> BlockMachine<T> {
 
         // Record the steps where we made progress, so we can report this to the
         // cache later on.
+        let mut progress_in_last_step = false;
         let mut progress_steps = vec![];
-        for step in sequence {
+
+        // We might or might not have a cached sequence. If not, we use the default one, which
+        // iterates over all identities until no more progress is made.
+        let mut cached_sequence_iter = cached_sequence.map(|s| s.into_iter());
+        let mut default_sequence_iter =
+            DefaultSequenceIterator::new(self.block_size, self.identities.len());
+
+        // Can't use a for loop here, because we need to communicate progress_in_last_step to the
+        // default iterator.
+        loop {
+            let step = match cached_sequence_iter.as_mut() {
```

Comment:
I'm not sure if this works out, but I would have imagined roughly the following architecture:
```
let mut iter = self.processing_sequence_cache.get_processing_sequence(left);
while let Some(step) = iter.next() {
  let mut progress_in_step = false;
  // do some stuff
  iter.report_progress(progress); 
}
```

I hope this can even be done in a way where we can remove the call to
```
            self.processing_sequence_cache
                .report_processing_sequence(left, progress_steps);
```
at the end.